sharding/proposer/service.go

```diff
@@ -43,42 +44,56 @@ func (p *Proposer) Start() {
 // Stop the main loop for proposing collations.
 func (p *Proposer) Stop() error {
 	log.Info(fmt.Sprintf("Stopping proposer service in shard %d", p.shardID))
+	p.txpoolSub.Unsubscribe()
 	return nil
 }
 
+// proposeCollations listens to the transaction feed submits a collation when one is found
 func (p *Proposer) proposeCollations() {
+	requests := make(chan *types.Transaction)
+	p.txpoolSub = p.txpool.TransactionsFeed().Subscribe(requests)
 
-	// TODO: Receive TXs from shard TX generator or TXpool (Github Issues 153 and 161)
-	var txs []*types.Transaction
-	for i := 0; i < 10; i++ {
-		data := make([]byte, 1024)
-		rand.Read(data)
-		txs = append(txs, types.NewTransaction(0, common.HexToAddress("0x0"),
-			nil, 0, nil, data))
+	for {
+		timeout := time.NewTimer(10 * time.Second)
```

Comment:
Please use context for timeouts
https://golang.org/pkg/context/#WithTimeout

Reply:
+1

Reply:
Ah, okay. Bear with me because I'm still new to Go. Are you saying call `context.WithTimeout` here and pass the new context down to `createCollation`, or to simply replace `time.NewTimer` with this context method and discard the context.

Reply:
Something like this:

```
for {
	ctx, cancel := context.WithTimeout(context.Background(), 10 * time.Second)
	defer cancel()

	select {
	case tx := <-requests:
		log.Info(fmt.Sprintf("Received transaction: %x", tx.Hash()))
		if err := p.createCollation(ctx, []*types.Transaction{tx}); err != nil {
			log.Error(fmt.Sprintf("Create collation failed: %v", err))
		}
	case <-ctx.Done():
		log.Error("Subscriber timed out")
	case err := <-p.txpoolSub.Err():
		log.Error(fmt.Sprintf("Subscriber failed: %v", err))
		cancel()
		return
	}
}
```